In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
acci = pd.read_csv('datasets\\accident_data.csv')

In [3]:
acci

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
acci.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [5]:
acci.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [6]:
acci.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [7]:
acci["Latitude"] = acci["Latitude"].fillna(acci["Latitude"].mean())
acci["Longitude"] = acci["Longitude"].fillna(acci["Longitude"].mean())
acci["Road_Surface_Conditions"] = acci["Road_Surface_Conditions"].fillna(acci["Road_Surface_Conditions"].mode()[0])
acci["Road_Type"] = acci["Road_Type"].fillna(acci["Road_Type"].mode()[0])
acci["Urban_or_Rural_Area"] = acci["Urban_or_Rural_Area"].fillna(acci["Urban_or_Rural_Area"].mode()[0])

In [8]:
acci.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [9]:
acci["Index"] = acci["Index"].astype('category')

In [10]:
acci["Accident_Severity"] = acci["Accident_Severity"].astype('category')

In [11]:
acci['Accident Date'] = pd.to_datetime(acci['Accident Date'] ,dayfirst = True, errors = 'coerce')

In [12]:
acci["Light_Conditions"] = acci["Light_Conditions"].astype('category')

In [13]:
acci["District Area"] = acci["District Area"].astype('category')

In [14]:
acci["Road_Surface_Conditions"] = acci["Road_Surface_Conditions"].astype('category')

In [15]:
acci["Road_Type"] = acci["Road_Type"].astype('category')

In [16]:
acci["Urban_or_Rural_Area"] = acci["Urban_or_Rural_Area"].astype('category')

In [17]:
acci["Weather_Conditions"] = acci["Weather_Conditions"].astype('category')

In [18]:
acci["Vehicle_Type"] = acci["Vehicle_Type"].astype('category')

In [19]:
acci['Year'] = acci['Accident Date'].dt.year
acci['Month'] = acci['Accident Date'].dt.month
acci['Day'] = acci['Accident Date'].dt.day
acci['DayOfWeek'] = acci['Accident Date'].dt.dayofweek


<h1 style="color: red;"> 20 PROBLEMS </h1>
<h2 style="color: BLUE;"> UNIVARIATE - Blue Insight</h2>
<h2 style="color: GREEN;"> BIVARIATE - Green Insight </h2>
<h2 style="color: VIOLET;"> MULTIVARIATE - Violet Insight </h2>


<h1>Problem 1 - What is the average number of casualties of accidents? </h1>

In [20]:
np.round(acci["Number_of_Casualties"].mean())

np.float64(1.0)

<h2 style="color: blue"> Insight no 1 - The average number of casualties is 1.0</h2>

<h1>Problem 2 - How many accidents were recorded in the dataset?</h1>

In [21]:
acci["Index"].count()

np.int64(660679)

<h2 style="color: blue"> Insight no 2 - The number of accidents recorded in datasets is 660,679</h2>

<h1>Problem 3 - What is the most car to drive under the wet or damp weather condition?</h1>

In [22]:
car_wet_safe  = acci.groupby(["Vehicle_Type","Road_Surface_Conditions"] ).size()

car_wet_safe.unstack()

Road_Surface_Conditions                   Dry  Flood over 3cm. deep  \
Vehicle_Type                                                          
Agricultural vehicle                     1304                     8   
Bus or coach (17 or more pass seats)    17623                    40   
Car                                    337860                   777   
Data missing or out of range                3                     0   
Goods 7.5 tonnes mgw and over           11708                    32   
Goods over 3.5t. and under 7.5t          4138                     6   
Minibus (8 - 16 passenger seats)         1356                     1   
Motorcycle 125cc and under              10515                    14   
Motorcycle 50cc and under                5200                    13   
Motorcycle over 125cc and up to 500cc    5233                     9   
Motorcycle over 500cc                   17465                    44   
Other vehicle                            3853                    11   
Pedal cycle                               138                     0   
Ridden horse                                3                     0   
Taxi/Private hire car                    9024                    13   
Van / Goods 3.5 tonnes mgw or under     23124                    49   

Road_Surface_Conditions                Frost or ice  Snow  Wet or damp  
Vehicle_Type                                                            
Agricultural vehicle                             59    24          552  
Bus or coach (17 or more pass seats)            646   214         7355  
Car                                           14108  4483       140764  
Data missing or out of range                      1     0            2  
Goods 7.5 tonnes mgw and over                   432   147         4988  
Goods over 3.5t. and under 7.5t                 162    51         1739  
Minibus (8 - 16 passenger seats)                 40    16          563  
Motorcycle 125cc and under                      420   128         4192  
Motorcycle 50cc and under                       224    77         2089  
Motorcycle over 125cc and up to 500cc           204    71         2139  
Motorcycle over 500cc                           689   226         7233  
Other vehicle                                   166    46         1561  
Pedal cycle                                      10     2           47  
Ridden horse                                      0     0            1  
Taxi/Private hire car                           384   122         3751  
Van / Goods 3.5 tonnes mgw or under             972   283         9732

<h2 style="color: green"> Insight no 3 - The safest car to drive under the wet or damp weather condition is Ridden Horse with accident rate of 1%</h2>

<h1>Problem 4 - How many unique road surface conditions are there? </h1>

In [23]:
acci["Road_Surface_Conditions"].nunique()

5

<h2 style="color: blue"> Insight no 4 - There are 5 unique road surface conditions in the dataset. Understanding these conditions helps in analyzing accident trends and identifying the most hazardous surfaces.</h2>

<h1>Problem 5 - What is the most common weather condition in the dataset?</h1>

In [24]:
acci["Weather_Conditions"].mode()[0]

'Fine no high winds'

<h2 style="color: blue"> Insight no 5 - "Fine no high winds" is the most common weather condition. This suggests that most accidents happen in clear weather, meaning human error and road conditions play a significant role.</h2>

<h1>Problem 6 - What is the Proportion of accidents in Urban vs. Rural areas?</h1>

In [25]:
accident_proportion_area = acci["Urban_or_Rural_Area"].value_counts(normalize=True).round(2) * 100
accident_proportion_area

Urban_or_Rural_Area
Urban          64.0
Rural          36.0
Unallocated     0.0
Name: proportion, dtype: float64

<h2 style="color: blue"> Insight no 6 - About 64% of accidents occur in urban areas, compared to 36% in rural areas. This is expected due to higher traffic volume, intersections, and pedestrian activity in urban settings.</h2>

<h1>Problem 7 - What is the  Most common accident severity in Urban vs. Rural areas?</h1>

In [26]:
most_common_severity_area = acci.groupby("Urban_or_Rural_Area")["Accident_Severity"].agg(lambda x: x.mode()[0])
most_common_severity_area

Urban_or_Rural_Area
Rural          Slight
Unallocated    Slight
Urban          Slight
Name: Accident_Severity, dtype: category
Categories (3, object): ['Fatal', 'Serious', 'Slight']

<h2 style="color: green"> Insight no 7 - The most common severity in both urban and rural areas is "Slight." Urban areas have more minor accidents, likely due to lower-speed collisions, while rural areas may have fewer but more severe accidents.</h2>

<h1>Problem 8 - What is the Most dangerous road surface condition based on total casualties?</h1>

In [27]:
casualties_by_road_surface = acci.groupby("Road_Surface_Conditions")["Number_of_Casualties"].sum().sort_values(ascending=False)
casualties_by_road_surface

Road_Surface_Conditions
Dry                     598606
Wet or damp             263457
Frost or ice             24976
Snow                      8008
Flood over 3cm. deep      1521
Name: Number_of_Casualties, dtype: int64

<h2 style="color: green"> Insight no 8 -  "Dry" roads have the highest number of casualties, but this is likely due to their frequent occurrence. However, wet and icy conditions show significantly higher accident severity.</h2>

<h1>Problem 9 - Weather condition with the highest accident count?</h1>

In [28]:
most_accidents_weather = acci["Weather_Conditions"].value_counts()
most_accidents_weather

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h2 style="color: blue"> Insight no 9 - The most common weather condition during accidents is "Fine no high winds". This suggests that most accidents occur under normal weather conditions, possibly due to higher traffic volume rather than hazardous weather.</h2>

<h1>Problem 10 - Which road type has the highest proportion of serious and fatal accidents?</h1>

In [29]:
serious_fatal_accidents = acci[acci["Accident_Severity"].isin(["Serious", "Fatal"])]
proportion_serious_fatal_road = (serious_fatal_accidents["Road_Type"].value_counts(normalize=True) * 100).round(2)

proportion_serious_fatal_road

Road_Type
Single carriageway    79.58
Dual carriageway      14.00
Roundabout             3.93
One way street         1.81
Slip road              0.68
Name: proportion, dtype: float64

<h2 style="color: violet"> Insight no 10 - Single carriageways have the highest proportion of serious and fatal accidents. This is likely due to high-speed limits, lack of dividers, and dangerous overtaking maneuvers.</h2>

<h1>Problem 11 - What is the average number of casualties in urban areas during daylight?</h1>

In [30]:
acci[(acci["Urban_or_Rural_Area"] == "Urban") & (acci["Light_Conditions"] == "Daylight")]["Number_of_Casualties"].mean()

np.float64(1.2688250554889653)

<h2 style="color: Violet"> Insight no 11 - The average number of casualties in urban daylight accidents is around 1.27. This suggests that while accidents are frequent, they typically involve only one or two casualties.</h2>

<h1>Problem 12 - What is the average number of casualties in rural areas at night?</h1>

In [31]:
acci[(acci["Urban_or_Rural_Area"] == "Rural") & (acci["Light_Conditions"].str.contains("Darkness"))]["Number_of_Casualties"].mean()


np.float64(1.514220615964802)

<h2 style="color: Violet"> Insight no 12 - The average casualties in rural nighttime accidents are around 1.51, which is slightly higher than urban accidents. This could be due to high-speed collisions and poor visibility at night.</h2>

<h1>Problem 13 - What is the most common vehicle type involved in slight accidents?</h1>

In [32]:
acci[acci["Accident_Severity"] == "Slight"]["Vehicle_Type"].mode()[0]


'Car'

<h2 style="color: Green"> Insight no 13 - Cars are the most common vehicle type involved in slight accidents. This is expected, as they make up the majority of road traffic.</h2>

<h1>Problem 14 - What is the most common road type for accidents in fine weather?</h1>

In [33]:
acci[acci["Weather_Conditions"] == "Fine no high winds"]["Road_Type"].mode()[0]


'Single carriageway'

<h2 style="color: Green"> Insight no 14 - TSingle carriageways are the most common road type for accidents in fine weather. This suggests that traffic volume, speed, and road design contribute more to accidents than weather conditions. </h2>

<h1>Problem 15 - What is the average number of vehicles involved in accidents when the weather is raining and it's night?</h1>

In [34]:
acci[(acci["Weather_Conditions"].str.contains("Raining", na=False)) & (acci["Light_Conditions"].str.contains("Darkness"))]["Number_of_Vehicles"].mean()


np.float64(1.7491552834493314)

<h2 style="color: Violet"> Insight no 15 - : On average, 1.75 vehicles are involved in nighttime rain accidents. This indicates a higher chance of multi-vehicle collisions, likely due to slippery roads and reduced visibility.</h2>

<h1>Problem 16 - Is there a correlation between the number of casualties and the number of vehicles involved?</h1>

In [35]:
casualties_vehicles_corr = np.round(acci["Number_of_Casualties"].corr(acci["Number_of_Vehicles"]), 1)
casualties_vehicles_corr

np.float64(0.2)

<h2 style="color: green"> Insight no 16 - There is a weak positive correlation (0.2), meaning that as the number of vehicles in an accident increases, the number of casualties tends to rise slightly.</h2>

<h1>Problem 17 - Is there a correlation between longitude and latitude?</h1>

In [36]:
location_corr = np.round(acci["Longitude"].corr(acci["Latitude"]), 1)
location_corr

np.float64(-0.4)

<h2 style="color: green"> Insight no 17 - There is a weak negative correlation (-0.4), which may indicate a geographical trend in accident locations but not a strong predictive factor.</h2>

<h1>Problem 18 - How many accident happen in the light condition "Day light" in 2019 </h1>

In [37]:
acci_2019 = acci[acci['Year'] == 2019]

day_light_sum = acci_2019["Light_Conditions"].value_counts()

day_light_sum

Light_Conditions
Daylight                       52518
Darkness - lights lit          14124
Darkness - no lighting          4242
Darkness - lighting unknown      719
Darkness - lights unlit          264
Name: count, dtype: int64

<h2 style="color: green"> Insight no 18 - In 2019, 52,518 accidents occurred in daylight. This highlights that most accidents happen during the day, possibly due to higher traffic density.</h2>

<h1>Problem 19 - What's the total of fatal severity accident that is caused by daylight</h1>

In [38]:
acci_sever = acci.groupby(["Accident_Severity","Light_Conditions"]).size()
acci_sever.unstack()

Light_Conditions   Darkness - lighting unknown  Darkness - lights lit  \
Accident_Severity                                                       
Fatal                                       68                   1860   
Serious                                    794                  19130   
Slight                                    5622                 108345   

Light_Conditions   Darkness - lights unlit  Darkness - no lighting  Daylight  
Accident_Severity                                                             
Fatal                                   45                    1612      5076  
Serious                                360                    7174     60759  
Slight                                2138                   28651    419045

<h2 style="color: green"> Insight no 19 - The number of fatal accidents in daylight is 5,076, which is the highest among all light conditions. This suggests that daylight does not necessarily reduce accident severity.</h2>

<h1>Problem 20 - What type of car is the highest accident rate</h1>

In [39]:
acci["Vehicle_Type"].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h2 style="color: blue"> Insight no 20 - Regular cars have the highest accident count, with nearly 500,000 accidents. This is expected since they are the most common type of vehicle on the road.</h2>

In [40]:
acci_2019 = acci[acci['Year'] == 2019]
acci_2020 = acci[acci['Year'] == 2020]
acci_2021 = acci[acci['Year'] == 2021]
acci_2022 = acci[acci['Year'] == 2022]

<h1>Problem 21 - What year has the highest accident rate?</h1>

In [41]:
acci["Year"].value_counts()

Year
2019.0    71867
2020.0    70163
2021.0    66172
2022.0    56805
Name: count, dtype: int64

<h2 style="color: blue"> Insight no 21 - The year 2019 has the highest accident rate recorded in datasets</h2>

<h1>Problem 22 - What year has the highest fatality? </h1>

In [42]:
acci_all_year = acci.groupby(["Accident_Severity","Year"]).size()
acci_all_year.unstack()

Year               2019.0  2020.0  2021.0  2022.0
Accident_Severity                                
Fatal                1116     942     823     623
Serious              9823    9346    8858    7266
Slight              60928   59875   56491   48916

<h2 style="color: green"> Insight no 22 - The year 2019 has the highest fatal rate recorded in datasets</h2>

<h1>Problem 23 - Does fatal accidents go down or up from the past records which is 2019 to the latest 2022?</h1>

In [43]:
acci_all_year = acci.groupby(["Accident_Severity","Year"]).size()
acci_all_year.unstack()

Year               2019.0  2020.0  2021.0  2022.0
Accident_Severity                                
Fatal                1116     942     823     623
Serious              9823    9346    8858    7266
Slight              60928   59875   56491   48916

<h2 style="color: green"> Insight no 23 - Based on the records, fatal accidents is going down as years goes by.</h2>

<h1>Problem 24 - What is the type of vehicles that has a high fatal record in 2019?</h1>

In [44]:
acci_fatal_2019 = acci[(acci['Year'] == 2019) & (acci['Accident_Severity'] == 'Fatal')]
acci_fatal_car_max = acci_fatal_2019.groupby(["Accident_Severity", "Vehicle_Type"]).size().idxmax()
acci_fatal_car = acci_fatal_2019.groupby(["Accident_Severity", "Vehicle_Type"]).size()
acci_fatal_car.unstack()


Vehicle_Type       Agricultural vehicle  Bus or coach (17 or more pass seats)  \
Accident_Severity                                                               
Fatal                                 2                                    64   
Serious                               0                                     0   
Slight                                0                                     0   

Vehicle_Type       Car  Data missing or out of range  \
Accident_Severity                                      
Fatal              798                             0   
Serious              0                             0   
Slight               0                             0   

Vehicle_Type       Goods 7.5 tonnes mgw and over  \
Accident_Severity                                  
Fatal                                         45   
Serious                                        0   
Slight                                         0   

Vehicle_Type       Goods over 3.5t. and under 7.5t  \
Accident_Severity                                    
Fatal                                           10   
Serious                                          0   
Slight                                           0   

Vehicle_Type       Minibus (8 - 16 passenger seats)  \
Accident_Severity                                     
Fatal                                             6   
Serious                                           0   
Slight                                            0   

Vehicle_Type       Motorcycle 125cc and under  Motorcycle 50cc and under  \
Accident_Severity                                                          
Fatal                                      37                         10   
Serious                                     0                          0   
Slight                                      0                          0   

Vehicle_Type       Motorcycle over 125cc and up to 500cc  \
Accident_Severity                                          
Fatal                                                 19   
Serious                                                0   
Slight                                                 0   

Vehicle_Type       Motorcycle over 500cc  Other vehicle  Pedal cycle  \
Accident_Severity                                                      
Fatal                                 38             10            2   
Serious                                0              0            0   
Slight                                 0              0            0   

Vehicle_Type       Ridden horse  Taxi/Private hire car  \
Accident_Severity                                        
Fatal                         0                     20   
Serious                       0                      0   
Slight                        0                      0   

Vehicle_Type       Van / Goods 3.5 tonnes mgw or under  
Accident_Severity                                       
Fatal                                               55  
Serious                                              0  
Slight                                               0

In [45]:
acci_fatal_car_max

('Fatal', 'Car')

<h2 style="color: green"> Insight no 24 - The type of vehicles that has a high fatal record in 2019 is a CAR with 71.5%</h2>